File: examples/example-kubernetes-client-openapi-informer-java/src/main/java/micronaut/client/configmap/ConfigMapInformer.java
package micronaut.client.configmap;

import io.micronaut.context.annotation.Context;
import io.micronaut.kubernetes.client.openapi.informer.SharedIndexInformer;
import io.micronaut.kubernetes.client.openapi.informer.SharedIndexInformerFactory;
import io.micronaut.kubernetes.client.openapi.informer.handler.ResourceEventHandler;
import io.micronaut.kubernetes.client.openapi.model.V1ConfigMap;
import jakarta.annotation.PostConstruct;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Context
public class ConfigMapInformer {

    private static final Logger LOG = LoggerFactory.getLogger(ConfigMapInformer.class);

    static final String NAMESPACE = "test-informer-namespace";

    private final SharedIndexInformerFactory sharedIndexInformerFactory;

    ConfigMapInformer(SharedIndexInformerFactory sharedIndexInformerFactory) {
        this.sharedIndexInformerFactory = sharedIndexInformerFactory;
    }

    @PostConstruct
    void initialize() {
        SharedIndexInformer<V1ConfigMap> sharedIndexInformer = sharedIndexInformerFactory.sharedIndexInformerFor(
            V1ConfigMap.class, NAMESPACE);
        sharedIndexInformer.addEventHandler(
            new ResourceEventHandler<>() {
                @Override
                public void onAdd(V1ConfigMap obj) {
                    LOG.info("{} config map added!", obj.getMetadata().getName());
                }

                @Override
                public void onUpdate(V1ConfigMap oldObj, V1ConfigMap newObj) {
                    LOG.info("{} config map updated!", oldObj.getMetadata().getName());
                }

                @Override
                public void onDelete(V1ConfigMap obj, boolean deletedFinalStateUnknown) {
                    LOG.info("{} config map deleted!", obj.getMetadata().getName());
                }
            });
    }
}

File: examples/example-kubernetes-client-openapi-informer-kotlin/src/main/kotlin/micronaut/client/configmap/ConfigMapInformer.kt
//tag::class[]
package micronaut.client.configmap

import io.micronaut.context.annotation.Context
import io.micronaut.kubernetes.client.openapi.informer.SharedIndexInformerFactory
import io.micronaut.kubernetes.client.openapi.informer.handler.ResourceEventHandler
import io.micronaut.kubernetes.client.openapi.model.V1ConfigMap
import io.micronaut.kubernetes.client.openapi.resolver.NamespaceResolver
import jakarta.annotation.PostConstruct
import org.slf4j.LoggerFactory
//end::class[]
import io.micronaut.context.annotation.Requires
import io.micronaut.context.env.Environment

@Requires(env = [Environment.KUBERNETES])
//tag::class[]

@Context
class ConfigMapInformer(
    private val sharedIndexInformerFactory: SharedIndexInformerFactory,
    private val namespaceResolver: NamespaceResolver
) {

    companion object {
        private val LOG = LoggerFactory.getLogger(ConfigMapInformer::class.java)
    }

    @PostConstruct
    fun initialize() {
        val namespace = namespaceResolver.resolveNamespace()
        val sharedIndexInformer = sharedIndexInformerFactory.sharedIndexInformerFor(
            V1ConfigMap::class.java, namespace)
        sharedIndexInformer.addEventHandler(
            object : ResourceEventHandler<V1ConfigMap> {
                override fun onAdd(obj: V1ConfigMap) {
                    LOG.info("{} config map added!", obj.metadata?.name)
                }

                override fun onUpdate(oldObj: V1ConfigMap, newObj: V1ConfigMap) {
                    LOG.info("{} config map updated!", oldObj.metadata?.name)
                }

                override fun onDelete(obj: V1ConfigMap, deletedFinalStateUnknown: Boolean) {
                    LOG.info("{} config map deleted!", obj.metadata?.name)
                }
            })
    }
}
//end::class[]
